features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java

```
@@ -0,0 +1,183 @@
+                if (resource == null) {
+                    continue;
+                }
+                if (resource.getResourceType().equals(Resource.Type.NODE)) {
+                    NodeResource nodeResource = (NodeResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(nodeResource.getNodeId());
+                    addAttributes(collectionSetResource, builder, nodeLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.INTERFACE)) {
+                    IpInterfaceResource ipResource = (IpInterfaceResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(ipResource.getNodeResource().getNodeId());
+                    InterfaceLevelResource interfaceLevelResource = new InterfaceLevelResource(nodeLevelResource, ipResource.getInstance());
+                    addAttributes(collectionSetResource, builder, interfaceLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.GENERIC)) {
+                    GenericTypeResource genericTypeResource = (GenericTypeResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(genericTypeResource.getNodeResource().getNodeId());
+                    // TODO: Get ResourceType from ResourceTypesDao
+                    org.opennms.netmgt.collection.support.builder.GenericTypeResource genericResource =
+                            new org.opennms.netmgt.collection.support.builder.GenericTypeResource(nodeLevelResource, null, genericTypeResource.getInstance());
+                    addAttributes(collectionSetResource, builder, genericResource);
+                }
+            }
+            builder.withTimestamp(new Date(collectionSet.getTimeStamp()));
+            return builder.build();
+
+        } catch (InterruptedException | ExecutionException e) {
+            LOG.error("collect failed with ", e);
```

We should probably rethrow this as a CollectionException instead of just returning `CollectionStatus.UNKNOWN`.